nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-core/src/main/java/org/apache/nifi/controller/status/history/storage/BufferedNodeStatusStorage.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.controller.status.history.storage;
+
+import org.apache.commons.math3.util.Pair;
+import org.apache.nifi.controller.status.NodeStatus;
+import org.apache.nifi.controller.status.history.StatusHistory;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * Decorator in front of a {@link NodeStatusStorage}. It accumulates entries to store within an internal buffer until
+ * method {@link #persist()} is being called, when buffer content is drained and being sent to te decorated instance.
+ */
+public class BufferedNodeStatusStorage implements NodeStatusStorage, Persistable {
+    private final BlockingQueue<Pair<Date, NodeStatus>> queue = new LinkedBlockingQueue<>();
+    private final NodeStatusStorage payload;
```

Comment:
`BufferedNodeStatusStorage` and `QuestDbNodeStatusStorage` implement this `NodeStatusStorage` interface but they cannot be used interchangeably. In `EmbeddedQuestDbNodeStatusRepository` a `BufferedNodeStatusStorage` needed (because it implements `Persistable` too) and here it does not really make sense to use a buffered instance (and it just would not work).
For this reason, it would be better to separate these two classes (eg. via removing `BufferedNodeStatusStorage` from the hierarchy).

Reply:
We had the same idea, I already pushed a change which makes the whole buffered (batched) writing somewhat more clear.